### PR TITLE
!!! TASK: Node name consistency adjustments

### LINKED ISSUE #4150: NodeAggregateNameWasChanged does not create copy on write
The [DoctrineDbalContentGraphProjection](https://github.com/neos/neos-development-collection/blob/003c8e65e3413af7e22172802ab3e0b4ddb9063e/Neos.ContentGraph.DoctrineDbalAdapter/src/DoctrineDbalContentGraphProjection.php#L380-L393) directly updates the hierarchyrelations without creating copies of the affected nodes first.

**Edit 12th April 2024:**
One of the issues is that the node timestamps are not updated when its name changed.
We decided to tackle this because it would require a projection replay if changed lateron.

### PR DESCRIPTION
**Pull request by nezaniel**
~draft until #4993 is merged~

builds on https://github.com/neos/neos-development-collection/pull/4993
resolves #4150 

This improves node name handling consistency by adjusting behavior in certain locations.

### Unique node names for children on aggregate level
Child node aggregates must no longer share a name, even if they don't share covered dimension space points.
This also means that `ContentGraphInterface::findChildNodeAggregateByName` now returns a single NodeAggregate, if any

Enforcement and respective tests have been adjusted

### Node names are now stored in the contentgraph's node table
Changing the node name now trigger a copy on write

### Manual structure adjustments regarding tethered nodes are now prohibited
Previously, it was possible to manually "fix" missing tethered nodes by renaming a node of the proper type or change the type of a node with the proper name. This is now no longer possible and instead to be performed globally via structure adjustments

**Upgrade instructions**

The method `ContentGraphInterface::findChildNodeAggregateByName` now returns a single NodeAggregate, if any and has been renamed:
```diff
- $contentGraph->findChildNodeAggregatesByName(...);
+ $contentGraph->findChildNodeAggregateByName(...);
```

`./flow cr:setup` and `./flow cr:replay contentgraph` is required due to node names now being stored in the node instead of hierarchy relation table

**Review instructions**

coming soon

**Checklist**

- [x] Code follows the PSR-12 coding style
- [x] Tests have been created, run and adjusted as needed
- [x] The PR is created against the 9.0 branch
- [ ] Reviewer - PR Title is brief but complete and starts with `FEATURE|TASK|BUGFIX`
- [ ] Reviewer - The first section explains the change briefly for change-logs
- [ ] Reviewer - Breaking Changes are marked with `!!!` and have upgrade-instructions

### CODE REVIEW

**Comment by mhsdesign** [comment]
Is this ready for review?

**Comment by nezaniel** [comment]
I'd like #5025 merged first, afterwards, yes, I think so

**Comment by nezaniel** [comment]
now open for review

**Comment by mhsdesign** [comment]
Oh no. This turns out to be in MAJOR conflict with @kitsunet month old https://github.com/neos/neos-development-collection/pull/5028 pr

- Neos.ContentGraph.DoctrineDbalAdapter/src/DoctrineDbalContentGraphProjection.php
- Neos.ContentGraph.DoctrineDbalAdapter/src/DoctrineDbalContentGraphSchemaBuilder.php
- Neos.ContentGraph.DoctrineDbalAdapter/src/Domain/Repository/ContentGraph.php
- Neos.ContentGraph.DoctrineDbalAdapter/src/Domain/Repository/ContentSubgraph.php
- Neos.ContentGraph.PostgreSQLAdapter/src/Domain/Repository/ContentHypergraph.php
- Neos.ContentRepository.Core/Classes/Feature/Common/ConstraintChecks.php
- Neos.ContentRepository.Core/Classes/Feature/Common/TetheredNodeInternals.php
- Neos.ContentRepository.Core/Classes/Feature/NodeCreation/NodeCreation.php
- Neos.ContentRepository.Core/Classes/Feature/NodeDuplication/NodeDuplicationCommandHandler.php
- Neos.ContentRepository.Core/Classes/Feature/NodeMove/NodeMove.php
- Neos.ContentRepository.Core/Classes/Feature/NodeRenaming/NodeRenaming.php
- Neos.ContentRepository.Core/Classes/Feature/NodeTypeChange/NodeTypeChange.php
- Neos.ContentRepository.Core/Classes/Projection/ContentGraph/ContentGraphInterface.php
- Neos.Neos/Classes/Controller/Module/Administration/SitesController.php
- Neos.Neos/Classes/Domain/Service/SiteServiceInternals.php

As there are also refactorings made regarding query building and stuff im not sure how to continue.
Maybe a compromise would be to remove these `NodeQueryBuilder` optimisations from your pr @kitsunet?
This pr fundamentally changes that the `name` of a node is now stored in the node table and not the hierarchy relation table.

Maybe even merge this one first as this contains real sql changes while the other one cosmetic ones and THESE changes must not get lost....

File: Neos.ContentGraph.DoctrineDbalAdapter/src/Domain/Projection/NodeRecord.php
Comment (mhsdesign) [inline Neos.ContentGraph.DoctrineDbalAdapter/src/Domain/Projection/NodeRecord.php:1]: First comment :D

The comment in line 43 will be out of date:

> Transient node name to store a node name after fetching a node with hierarchy (not always available)

Comment (mhsdesign) [reply Neos.ContentGraph.DoctrineDbalAdapter/src/Domain/Projection/NodeRecord.php:1]: And also out of date is this comment https://github.com/neos/neos-development-collection/blob/7977bac7dc7fb8e4bf1f7accb5668d5e005fa69b/Neos.ContentRepository.Core/Classes/Projection/ContentGraph/Node.php#L49

> @param NodeName|null $nodeName The optionally named hierarchy relation to the node's parent.

Comment (nezaniel) [reply Neos.ContentGraph.DoctrineDbalAdapter/src/Domain/Projection/NodeRecord.php:1]: true thing, fixed :)

**Comment by nezaniel** [comment]
Wasn't that bad after all. I'd like #5040 merged first, though, to resolve the expected additional conflicts here

File: Neos.ContentGraph.DoctrineDbalAdapter/src/DoctrineDbalContentGraphSchemaBuilder.php [hunk 1]
```diff
@@ -39,6 +39,7 @@ private function createNodeTable(): Table
             DbalSchemaFactory::columnForNodeAggregateId('nodeaggregateid')->setNotnull(false),
             DbalSchemaFactory::columnForDimensionSpacePointHash('origindimensionspacepointhash')->setNotnull(false),
             DbalSchemaFactory::columnForNodeTypeName('nodetypename'),
+            (new Column('name', Type::getType(Types::STRING)))->setLength(255)->setNotnull(false)->setCustomSchemaOption('charset', 'ascii')->setCustomSchemaOption('collation', 'ascii_general_ci'),
```
Comment (mhsdesign) [inline Neos.ContentGraph.DoctrineDbalAdapter/src/DoctrineDbalContentGraphSchemaBuilder.php:42]: this should probably be reviewed by someone who knows better than me :D

File: Neos.ContentRepository.Core/Classes/Feature/Common/ConstraintChecks.php [hunk 2]
```diff
@@ -324,37 +327,16 @@ protected function requireNodeTypeConstraintsImposedByParentToBeMet(
...
-            && !$this->getNodeTypeManager()->getTypeOfTetheredNode($parentsNodeType, $nodeName)->name->equals($nodeType->name)
-        ) {
-            throw new NodeConstraintException(
-                'Node type "' . $nodeType->name->value . '" does not match configured "'
-                    . $this->getNodeTypeManager()->getTypeOfTetheredNode($parentsNodeType, $nodeName)->name->value
-                    . '" for auto created child nodes for parent type "' . $parentsNodeType->name->value
-                    . '" with name "' . $nodeName->value . '"',
-                1707561404
-            );
-        }
```
Comment (mhsdesign) [inline Neos.ContentRepository.Core/Classes/Feature/Common/ConstraintChecks.php:339]: why were these constraint checks removed? I came across those many times and last time i read them they seemed correct?

File: Neos.ContentRepository.Core/Classes/Feature/Common/ConstraintChecks.php [hunk 3]
```diff
@@ -602,63 +584,27 @@ protected function requireNodeAggregateToBeChild(
         );
     }
 
-    /**
-     * @throws NodeNameIsAlreadyOccupied
-     */
-    protected function requireNodeNameToBeUnoccupied(
```
Comment (mhsdesign) [inline Neos.ContentRepository.Core/Classes/Feature/Common/ConstraintChecks.php:608]: a i see thats what you meant, we exchange this check for the simpler, non dimension aware `requireNodeNameToBeUncovered` thus this method can be removed and its callers rewritten

File: Neos.ContentRepository.Core/Classes/Feature/Common/TetheredNodeInternals.php [hunk 4]
```diff
@@ -152,11 +143,6 @@ protected function createEventsForMissingTetheredNode(
                 $originDimensionSpacePoint,
                 $parentNodeAggregate
             );
-        } else {
-            throw new \RuntimeException(
-                'There is >= 2 ChildNodeAggregates with the same name reachable from the parent' .
-                    '- this is ambiguous and we should analyze how this may happen. That is very likely a bug.'
-            );
```
Comment (mhsdesign) [inline Neos.ContentRepository.Core/Classes/Feature/Common/TetheredNodeInternals.php:159]: nice :D

File: Neos.Neos/Classes/Controller/Module/Administration/SitesController.php [hunk 5]
```diff
@@ -204,15 +204,11 @@ public function updateSiteAction(Site $site, $newSiteNodeName)
             }
 
             foreach ($contentRepository->getWorkspaceFinder()->findAll() as $workspace) {
-                // technically, due to the name being the "identifier", there might be more than one :/
```
Comment (mhsdesign) [inline Neos.Neos/Classes/Controller/Module/Administration/SitesController.php:207]: thanks for removing this mindfuck :D

**Review by mhsdesign — APPROVED** [review]
Looks good already ;) Thanks for including the adjustments!

The Neos Ui e2e also pass without any adjustments
(as we want to release a new beta soon today tomorrow? i would not want to include it into that because of the replay, but it can be merged the second the beta was released)

File: Neos.ContentGraph.DoctrineDbalAdapter/src/DoctrineDbalContentGraphSchemaBuilder.php (hunk 1 above)
Comment (nezaniel) [reply Neos.ContentGraph.DoctrineDbalAdapter/src/DoctrineDbalContentGraphSchemaBuilder.php:42]: that's just cut&paste from the hierarchy relation table ^^

File: Neos.ContentRepository.Core/Classes/Feature/Common/ConstraintChecks.php (hunk 2 above)
Comment (nezaniel) [reply Neos.ContentRepository.Core/Classes/Feature/Common/ConstraintChecks.php:339]: They were replaced by an even stricter  requireNodeTypeNotToDeclareTetheredChildNodeName (see above)

File: Neos.ContentRepository.Core/Classes/Feature/Common/ConstraintChecks.php (hunk 3 above)
Comment (nezaniel) [reply Neos.ContentRepository.Core/Classes/Feature/Common/ConstraintChecks.php:608]: exactly, this was only needed for checking that if two node aggregates with the same name share a parent, they must not share dimension space points. Which is no longer allowed in the first place.

Comment (mhsdesign) [reply Neos.ContentRepository.Core/Classes/Feature/Common/ConstraintChecks.php:608]: Okay i see but my feeling says `requireNodeTypeNotToDeclareTetheredChildNodeName` is not called often enough? Only once in the Move handling?

File: Neos.ContentRepository.Core/Classes/Feature/Common/ConstraintChecks.php (hunk 2 above)
Comment (mhsdesign) [reply Neos.ContentRepository.Core/Classes/Feature/Common/ConstraintChecks.php:339]: Okay i see but my feeling says `requireNodeTypeNotToDeclareTetheredChildNodeName` is not called often enough? Only once in the Move handling?

Comment (nezaniel) [reply Neos.ContentRepository.Core/Classes/Feature/Common/ConstraintChecks.php:339]: It's called three times, at all relevant locations:
* NodeCreation: You can't manually create a missing tethered node anymore
* NodeMove: You can't move a node to a parent where a tethered child of the same name is missing
* NodeRenamig: You can't assign the name of a missing tethered sibling to a node anymore

**Comment by nezaniel** [comment]
> (as we want to release a new beta soon today tomorrow? i would not want to include it into that because of the replay, but it can be merged the second the beta was released)

Is there any particular reason why the replay should be in the next beta instead of this one?

**Comment by mhsdesign** [comment]
idk but it seems no replay for this beta is not possible see: https://github.com/neos/neos-development-collection/pull/5025#issuecomment-2106256549

was just thinking about keeping things simple when beta hopping and putting all migrate related stuff in one batch but youre rm as well so was just a slight thought

**Comment by nezaniel** [comment]
As this has to be merged (rather sooner than later) anyway, I'd say we can do so already

File: Neos.ContentRepository.Core/Classes/Feature/Common/TetheredNodeInternals.php (hunk 4 above)
Comment (nezaniel) [reply Neos.ContentRepository.Core/Classes/Feature/Common/TetheredNodeInternals.php:159]: The pleasure was all mine

File: Neos.Neos/Classes/Controller/Module/Administration/SitesController.php (hunk 5 above)
Comment (nezaniel) [reply Neos.Neos/Classes/Controller/Module/Administration/SitesController.php:207]: You're welcome 🙂

File: Neos.ContentRepository.Core/Classes/Feature/Common/ConstraintChecks.php (hunk 2 above)
Comment (mhsdesign) [reply Neos.ContentRepository.Core/Classes/Feature/Common/ConstraintChecks.php:339]: Wait but what about `handleCopyNodesRecursively` and `handleChangeNodeAggregateType`? they had a call previously to `requireConstraintsImposedByAncestorsAreMet` but not to `requireNodeTypeNotToDeclareTetheredChildNodeName`?

Comment (nezaniel) [reply Neos.ContentRepository.Core/Classes/Feature/Common/ConstraintChecks.php:339]: Good point, I'll have to check what copy does with names.
Type change is unrelated though as it does not affect the name

**Review by kitsunet — APPROVED** [review]
👀

**Pull request merged**